src/runtime_src/core/common/api/xrt_module.cpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (C) 2023 Advanced Micro Devices, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+#define XCL_DRIVER_DLL_EXPORT  // exporting xrt_module.h
+#define XRT_API_SOURCE         // exporting xrt_module.h
+#define XRT_CORE_COMMON_SOURCE // in same dll as core_common
+#include "experimental/xrt_module.h"
+#include "experimental/xrt_elf.h"
+
+#include "xrt/xrt_bo.h"
+#include "xrt/xrt_hw_context.h"
+#include "xrt/xrt_uuid.h"
+
+#include "module_int.h"
+#include "core/common/debug.h"
+#include "core/common/error.h"
+
+#include <cstring>
+#include <string>
+
+namespace {
+
+} // namespace
+
+namespace xrt {
+
+// class module_impl - Base class for different implementations
+class module_impl
+{
+protected:
+  xrt::uuid m_cfg_uuid;     // matching hw configuration id
+
+public:
+  module_impl(xrt::uuid cfg_uuid)
+    : m_cfg_uuid(std::move(cfg_uuid))
+  {}
+
+  module_impl(const module_impl* parent)
+    : m_cfg_uuid(parent->m_cfg_uuid)
+  {}
+
+  virtual
+  ~module_impl()
+  {}
+
+  xrt::uuid
+  get_cfg_uuid() const
+  {
+    return m_cfg_uuid;
+  }
+
+  virtual std::pair<const char*, size_t>
+  get_data() const
+  {
+    return {nullptr, 0};
+  }
+
+  virtual xrt::hw_context
+  get_hw_context() const
+  {
+    return {};
+  }
+
+  virtual xrt::bo
+  get_instruction_buffer(const std::string&) const
+  {
+    return {};
+  }
+};
+
+// class module_elf - Elf provided by application
+class module_elf : public module_impl
+{
+  xrt::elf m_elf;
+
+public:
+  module_elf(xrt::elf elf)
+    : module_impl{elf.get_cfg_uuid()}
+    , m_elf(std::move(elf))
+  {}
+
+  virtual std::pair<const char*, size_t>
+  get_data() const override
+  {
+    return {nullptr, 0}; // TBD
+  }
+};
+
+// class module_userptr - Opaque userptr provided by application
+class module_userptr : public module_impl
+{
+  std::vector<char> m_buffer; // userptr copy
+
+public:
+  module_userptr(char* userptr, size_t sz, const xrt::uuid& uuid)
+    : module_impl{uuid}
+    , m_buffer{userptr, userptr + sz}
+  {}
+
+  module_userptr(void* userptr, size_t sz, const xrt::uuid& uuid)
+    : module_userptr(static_cast<char*>(userptr), sz, uuid)
+  {}
+
+  virtual std::pair<const char*, size_t>
+  get_data() const override
+  {
+    return {m_buffer.data(), m_buffer.size()};
+  }
+};
+
+// class module_sram - Create an sram module from parent
+class module_sram : public module_impl
```

Comment:
This is only applicable to IPU PHX and STX. There is no SRAM on Turin. Do we want a generic name? Since this is common code, we can support xrt::bo::flags::cacheable flag and malloc normal memory in shim on Turin. The downside would be that we have to make a copy of instruction buffer even on Turin....

Reply:
Yeah, it's just a name, completely internal to this one .cpp file.   We can call it whatever we like and we can change the name whenever we want.   We also need to change the "cacheable" flag name, it makes no sense.  This code will change as we adopt it,  I am not changing at this point.

Reply:
Sounds good.